src/transformStateWithClones.js

```diff
@@ -1,13 +1,39 @@
-'use strict';
+const ACTION_TYPE_CLEAR = 'clear';
+const ACTION_TYPE_ADD_PROPERTIES = 'addProperties';
+const ACTION_TYPE_REMOVE_PROPERTIES = 'removeProperties';
 
 /**
- * @param {Object} state
- * @param {Object[]} actions
+ * Оновлює стан об'єкта на основі масиву дій і повертає історію змін.
  *
- * @return {Object[]}
+ * @param {Object} initialState - Початковий стан об'єкта.
+ * @param {Object[]} actions - Масив дій, які будуть застосовані до стану.
+ * @returns {Object[]} - Масив станів після кожної дії.
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(initialState, actions) {
+  const stateHistory = [];
+  let currentState = { ...initialState };
+
+  actions.forEach((action) => {
+    if (action.type === ACTION_TYPE_CLEAR) {
+      currentState = {};
+    }
+
+    if (action.type === ACTION_TYPE_ADD_PROPERTIES) {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === ACTION_TYPE_REMOVE_PROPERTIES) {
+      currentState = { ...currentState };
+
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    }
+
+    stateHistory.push(currentState);
```

Comment:
Instead of using multiple `if` statements to handle action types, consider using a `switch` statement with a `default` case for error handling. This aligns with the [CODE STYLE] checklist item that suggests using a `switch` statement when you have a limited number of conditions, and it also ensures that you handle unexpected action types. Furthermore, the `default` case can be used to throw an error or handle unrecognized action types.